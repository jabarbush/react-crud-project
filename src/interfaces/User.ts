interface User {
    id: number;
    name: string;
    dob: string;
    phone: string;
    email: string;
    signature?: {
        pin: string;
        fontStyle: string;
    } | null;
  }
  
  export default User;
  