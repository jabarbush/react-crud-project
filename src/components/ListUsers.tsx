import React, { useState } from 'react';
import User from '../interfaces/User';
import ListProps from '../interfaces/ListProps';
import UserInfo from './UserInfo';
import { ReactComponent as PanelLeftClose } from '../assets/panel-left-close.svg';

const List: React.FC<ListProps> = ({ userList, setUserList }) => {
  const [selectedUser, setSelectedUser] = useState<User | null>(null);

  const handleUserClick = (user: User) => {
    setSelectedUser(user);
  };

  const handleCloseUserInfo = () => {
    setSelectedUser(null);
  };

  const UserInfoActive = selectedUser !== null;

  return (
    <div>
      {UserInfoActive && (
        <UserInfo userList={userList} selectedUser={selectedUser} onCloseUserInfo={handleCloseUserInfo} setUserList={setUserList} />
      )}
      <div className={`list-container ${UserInfoActive ? 'active' : ''}`}>
        {userList.map((user) => (
          <div key={user.id} className={`card ${selectedUser === user ? 'active' : ''}`}>
            <div>
              <div className="user-name">{user.name}</div>
              <div className="dob-card">DOB: {user.dob}</div>
            </div>
            <div className={`signature-status ${user.signature ? 'green' : 'red'}`}>
              {user.signature ? 'Has Custom Signature' : 'Missing Custom Signature'}
            </div>
            <PanelLeftClose onClick={() => handleUserClick(user)} className='panel-icon' style={{ color: selectedUser === user ? '#1890ff' : '' }}/>
          </div>
        ))}
      </div>
    </div>
  );
};

export default List;
