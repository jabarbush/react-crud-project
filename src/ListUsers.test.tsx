import { render, screen } from '@testing-library/react';
import User from './interfaces/User';
import ListUsers from './components/ListUsers';
import { SetStateAction } from 'react';

beforeEach(() => jest.resetAllMocks())
describe('ListUsers component', () => {
    const selectedUser = {
        id: 1,
        name: 'jake',
        dob: '03/03/2003',
        phone: '4564564564',
        email: 'jake4@gmail.com',
        signature: {
          pin: '123123',
          fontStyle: 'Mr Dafoe',
        },
      };
    const userList: User[] = [{
        id: 1,
        name: 'jake',
        dob: '03/03/2003',
        phone: '4564564564',
        email: 'jake4@gmail.com',
        signature: {
            pin: '123123',
            fontStyle: 'Mr Dafoe',
        },
    },
    {
        "id": 2,
        "name": "jack425525333",
        "dob": "03/03/2003",
        "phone": "4564564564",
        "email": "jake4@gmail.com",
        "signature": null
      },
      {
        "id": 3,
        "name": "ashley",
        "dob": "05/05/1995",
        "phone": "0980980988",
        "email": "ashley@jake.com",
        "signature": {
          "pin": "123123",
          "fontStyle": "Dancing Script"
        }
      },
      {
        "id": 4,
        "name": "123",
        "dob": "01/01/2001",
        "phone": "1231231231",
        "email": "jake@gmail.com",
        "signature": {
          "pin": "123123",
          "fontStyle": "Great Vibes"
        }
      }];

    const setUserList = jest.fn();
    const onUserClick = jest.fn();

    test('ListUsers component renders', () => {
        render(
            <ListUsers
                selectedUser={selectedUser}
                userList={userList}
                onUserClick={onUserClick} setUserList={function (value: SetStateAction<User[]>): void {
                    throw new Error('Function not implemented.');
                } }            />
        );
        const list = screen.getByTestId('list-component');
        expect(list).toBeInTheDocument();
    });

    test('ListUsers component renders each user in list', () => {
        render(
            <ListUsers
                selectedUser={selectedUser}
                userList={userList}
                onUserClick={onUserClick} setUserList={function (value: SetStateAction<User[]>): void {
                    throw new Error('Function not implemented.');
                } }            />
        );
        const user1 = screen.getByText('jake');
        const user2 = screen.getByText('jack425525333');
        const user3 = screen.getByText('ashley');
        const user4 = screen.getByText('123');

        expect(user1).toBeInTheDocument();
        expect(user2).toBeInTheDocument();
        expect(user3).toBeInTheDocument();
        expect(user4).toBeInTheDocument();
    });

    test('ListUsers component matches snapshot', () => {
      
        const { asFragment } = render(
            <ListUsers
                selectedUser={selectedUser}
                setUserList={setUserList}
                onUserClick={onUserClick} userList={[]}
            />
        );
      
        expect(asFragment()).toMatchSnapshot();
      });
});