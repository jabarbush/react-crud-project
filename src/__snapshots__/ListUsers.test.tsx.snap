// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListUsers component ListUsers component matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="list-component"
  >
    <div
      class="list-container"
      data-testid="user-item"
    />
  </div>
</DocumentFragment>
`;
